### PR TITLE
TLS v1.2 and v1.3 RSA PSS

### PR DESCRIPTION
**Pull request by SparkiDev**
Cleanup the TLS v1.3 PSS code as well.
Added RSA API wc_RsaPSS_CheckPadding() to check the padding - no longer
a simple memcmp with the digest.

### CODE REVIEW

File: src/tls.c [hunk 1]
```diff
@@ -4420,66 +4420,28 @@ static int TLSX_SetSupportedVersions(TLSX** extensions, const void* data,
...
-#endif
+static void TLSX_SignatureAlgorithms_MapPss(WOLFSSL *ssl, byte* input,
+                                            word16 length)
+{
+    word16 i;
 
-    return OPAQUE16_LEN + cnt * OPAQUE16_LEN;
+    ssl->pssAlgo = 0;
+    for (i = 0; i < length; i += 2) {
+        if (input[i] == rsa_pss_sa_algo && input[i + 1] <= sha512_mac)
```
Comment (toddouska) [inline src/tls.c]: Accessing i+1 does not look valid because i is compared to length with <.

File: wolfcrypt/src/rsa.c [hunk 2]
```diff
@@ -1577,6 +1586,30 @@ int wc_RsaPSS_VerifyInline(byte* in, word32 inLen, byte** out,
...
+{
+    int ret;
+
+    if (in == NULL || sig == NULL ||
+                      inSz != (word32)wc_HashGetDigestSize(hashType) ||
+                      sigSz != RSA_PSS_PAD_SZ + inSz * 3)
+        ret = BAD_FUNC_ARG;
+    else {
+        XMEMCPY(sig + RSA_PSS_PAD_SZ, in, inSz);
+        wc_Hash(hashType, sig, RSA_PSS_PAD_SZ + inSz * 2, sig, inSz);
```
Comment (toddouska) [inline wolfcrypt/src/rsa.c]: Why isn't the return code checked?

File: wolfcrypt/src/rsa.c [hunk 3]
```diff
@@ -613,7 +617,8 @@ static int RsaPad_PSS(const byte* input, word32 inputLen, byte* pkcsBlock,
     ret = RsaMGF(mgf, h, hLen, pkcsBlock, pkcsBlockLen - hLen - 1, heap);
     if (ret != 0)
         return ret;
-    pkcsBlock[0] &= 0x7f;
+    /* TODO: use the number of bits in the prime */
```
Comment (toddouska) [inline wolfcrypt/src/rsa.c]: What are the ramifications of not using the correct bits here?

File: src/tls.c (hunk 1 above)
Comment (SparkiDev) [inline src/tls.c]: The data is internal, when TLSX_SignatureAlgorithms_Write() calls the function, or external, when TLSX_SignatureAlgorithms_Parse() calls the funciton.
The internal data is always assumed to be valid (even number of bytes)
The external data is checked to ensure that the length of data is even in TLSX_SignatureAlgorithms_Parse().

File: wolfcrypt/src/rsa.c (hunk 3 above)
Comment (SparkiDev) [inline wolfcrypt/src/rsa.c]: Message will not verify if the number of bits is wrong.
Added code to fix this.

File: wolfcrypt/src/rsa.c (hunk 2 above)
Comment (SparkiDev) [inline wolfcrypt/src/rsa.c]: Return now checked.

File: src/tls.c (hunk 1 above)
Comment (toddouska) [inline src/tls.c]: We can't rely on side effects that may change in the future that aren't documented and easily enforced.  If we're going to access i+1 in a loop we should validate i+1 is legal to access.

Comment (SparkiDev) [inline src/tls.c]: Function now checks and returns an error.

**Pull request merged**